core/platform/email/gae_email_services_test.py

```diff
@@ -43,3 +49,132 @@ def test_sending_email(self):
+    SENDER_USERNAME = 'sender'
+    RECIPIENT_A_EMAIL = 'a@example.com'
+    RECIPIENT_A_USERNAME = 'usera'
+    RECIPIENT_B_EMAIL = 'b@example.com'
+    RECIPIENT_B_USERNAME = 'userb'
+    RECIPIENT_EMAILS = [RECIPIENT_A_EMAIL, RECIPIENT_B_EMAIL]
+
+    def setUp(self):
+        super(BulkEmailsTests, self).setUp()
+        # SENDER is authorised sender.
+        # A and B are recipients.
+        self.signup(self.SENDER_EMAIL, self.SENDER_USERNAME)
+        self.signup(self.RECIPIENT_A_EMAIL, self.RECIPIENT_A_USERNAME)
+        self.signup(self.RECIPIENT_B_EMAIL, self.RECIPIENT_B_USERNAME)
+        self.recipient_emails = [self.RECIPIENT_A_EMAIL, self.RECIPIENT_B_EMAIL]
+
+    def test_that_correct_bulk_emails_sent(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL,
+                self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+            messages_a = self.mail_stub.get_sent_messages(
+                to=self.RECIPIENT_A_EMAIL)
+            self.assertEqual(len(messages_a), 1)
+            messages_b = self.mail_stub.get_sent_messages(
+                to=self.RECIPIENT_B_EMAIL)
+            self.assertEqual(len(messages_b), 1)
+
+    def test_that_email_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        email_exception = (
+            self.assertRaisesRegexp(
+                Exception,
+                'This app cannot send emails.'))
+        with email_exception:
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.recipient_emails,
+                'subject', 'body', 'html')
+            messages_a = self.mail_stub.get_sent_messages(
+                to=self.RECIPIENT_A_EMAIL)
+            self.assertEqual(len(messages_a), 0)
+
+            messages_b = self.mail_stub.get_sent_messages(
+                to=self.RECIPIENT_B_EMAIL)
+            self.assertEqual(len(messages_b), 0)
+
+    def test_that_bulk_mails_not_sent_if_sender_email_is_malformed(self):
+        malformed_sender_email = ''
+        email_exception = (
+            self.assertRaisesRegexp(
+                ValueError,
+                'Malformed sender email address: %s'
+                % malformed_sender_email)
+        )
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_bulk_mail(
+                malformed_sender_email, self.recipient_emails,
+                'subject', 'body', 'html')
+            messages_a = self.mail_stub.get_sent_messages(
+                to=self.RECIPIENT_A_EMAIL)
+            self.assertEqual(0, len(messages_a))
+            messages_b = self.mail_stub.get_sent_messages(
+                to=self.RECIPIENT_B_EMAIL)
+            self.assertEqual(0, len(messages_b))
+
+    def test_that_bulk_mails_not_sent_if_recepient_email_is_malformed(self):
```

Comment:
Think of other tests. For example, what if one recipient email is malformed and one isn't? Try and ensure that the tests specify fully the behaviour of the code (don't just think about coverage only).

Reply:
Added other tests too. Thanks!